### PR TITLE
Align with stactools

### LINKED ISSUE #155: Follow STAC best practices
**Roadmap to move `ukis_pysat.data` to STAC**

- [x] Return metadata as pystac Items.
- [x] Develop our own STAC-API search class (inspired by [sat-search](https://github.com/sat-utils/sat-search/blob/master/satsearch/search.py)).
- [ ] Convert metadata items returned by data providers dynamically to [STAC Items according to best practises for different platforms](https://github.com/stac-utils/stac-sentinel). Currently we do this hardcoded and only for a small subset of attributes. The full set of attributes should also include the image band list.

_Originally posted by @MWieland in https://github.com/dlr-eoc/ukis-pysat/issues/23#issuecomment-862729500_

Could be based on these [packages](https://github.com/stactools-packages).

### PR DESCRIPTION
**Pull request by fwfichtner**
Closes #155 

It is not perfect, but the alignment is better. Some things can just not be known before having the actual files.

### CODE REVIEW

File: ukis_pysat/data.py [hunk 1]
```diff
@@ -254,22 +258,33 @@ def construct_metadata(self, meta, platform):
...
 
+            eo_ext = EOExtension.ext(item, add_if_missing=True)
+            if "cloudcoverpercentage" in meta["properties"]:
+                eo_ext.cloud_cover = round(float(meta["properties"]["cloudcoverpercentage"]), 2)
+
+            if platform == platform.Sentinel1:
+                utils.fill_sar_ext(SarExtension.ext(item, add_if_missing=True), meta=meta)
+
+            # common
+            item.common_metadata.providers = [SENTINEL_PROVIDER]
```
Comment (MWieland) [inline ukis_pysat/data.py:284]: The provider attribute for earthexplorer is missing at the moment.

File: ukis_pysat/utils.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    # Read properties
+    sar_ext.instrument_mode = meta["properties"]["sensoroperationalmode"]
+    sar_ext.product_type = meta["properties"]["producttype"]
+
+    # TODO maybe this is not good, because we often only use one later on
+    sar_ext.polarizations = [Polarization(p) for p in meta["properties"]["polarisationmode"].split(" ")]
+
+
+# constants
+SENTINEL_PROVIDER = pystac.Provider(
```
Comment (MWieland) [inline ukis_pysat/utils.py:22]: Suggest to add Landsat provider as well.

Comment (MWieland) [reply ukis_pysat/utils.py:22]:
```python
LANDSAT_PROVIDER = pystac.Provider(    
    name="USGS",
    roles=[
        ProviderRole.PRODUCER,
        ProviderRole.PROCESSOR,
        ProviderRole.LICENSOR,
    ],
    url="https://earthexplorer.usgs.gov/",
)
```

File: ukis_pysat/utils.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import pystac
+from geojson import Feature
+from pystac import ProviderRole
+from pystac.extensions.sar import FrequencyBand, SarExtension, Polarization
+
+
+def fill_sar_ext(sar_ext: SarExtension, meta: Feature):
```
Comment (MWieland) [inline ukis_pysat/utils.py:7]: do we really need/want this one? if we would stay consistent throughout the different platforms, we should also consider to use `eo:bands` for the multi-spectral platforms. also i am not so happy with the current placement of `producttype` inside the sar_ext. imu this should be a more general attribute and is not SAR specific. The discussion on this topic seems also not fully converged yet at specs level: https://github.com/radiantearth/stac-spec/issues/897

File: ukis_pysat/members.py [hunk 4]
```diff
@@ -1,11 +1,12 @@
 from enum import Enum
+
 from pydantic import BaseModel
 
 
 class Platform(Enum):
     Sentinel1 = "Sentinel-1"
     Sentinel2 = "Sentinel-2"
-    Sentinel3 = "Sentinel-3"
+
```
Comment (MWieland) [inline ukis_pysat/members.py:9]: I don't clearly see why we need to drop Sentinel-3.

**Review by MWieland — COMMENTED** [review]
A very good step forward, just a couple of points for discussion.

File: ukis_pysat/data.py [hunk 5]
```diff
@@ -254,22 +258,33 @@ def construct_metadata(self, meta, platform):
...
-            sat_ext = SatExtension.ext(item)
-
-            if "cloudcoverpercentage" in meta["properties"]:
-                eo_ext.cloud_cover = round(float(meta["properties"]["cloudcoverpercentage"]), 2)
 
-            item.common_metadata.platform = platform.value
+            # extensions
+            proj_ext = ProjectionExtension.ext(item, add_if_missing=True)
+            proj_ext.apply(
+                epsg=4326,
```
[the remark this answers is not in the document]
Comment (MWieland) [reply ukis_pysat/data.py:267]: Perhaps you meant this with the TODO in line 231? If so, I think the epsg code is part of the original metadata from the datahubs and we could use it. Not sure though - need to check this.

File: ukis_pysat/data.py [hunk 6]
```diff
@@ -223,19 +226,20 @@ def construct_metadata(self, meta, platform):
                     "end_datetime": meta["stop_time"].astimezone(tz=datetime.timezone.utc).isoformat(),
                 },
             )
-            EOExtension.add_to(item)
-            SatExtension.add_to(item)
-            eo_ext = EOExtension.ext(item)
-            sat_ext = SatExtension.ext(item)
+            # extensions
+            # proj_ext = ProjectionExtension.ext(item, add_if_missing=True)
+            # TODO can we know anything?
```
[the remark this answers is not in the document]
Comment (fwfichtner) [reply ukis_pysat/data.py:231]: At least in what we recorded I did not find this information

File: ukis_pysat/utils.py (hunk 2 above)
Comment (fwfichtner) [reply ukis_pysat/utils.py:22]: It was not in [here](https://github.com/stactools-packages/landsat).

File: ukis_pysat/data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fwfichtner) [reply ukis_pysat/data.py:268]: I just took it from [there](https://github.com/stactools-packages/sentinel1-grd/blob/main/src/stactools/sentinel1_grd/properties.py#L80). For Sentinel-2 it's read from the [manifest](https://github.com/stactools-packages/sentinel2/blob/main/src/stactools/sentinel2/stac.py#L47), for Sentinel-3 it is not set.

File: ukis_pysat/members.py (hunk 4 above)
Comment (fwfichtner) [reply ukis_pysat/members.py:9]: If we remove things like https://github.com/dlr-eoc/ukis-pysat/pull/163#discussion_r822037081 we can keep it I guess.

File: ukis_pysat/data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fwfichtner) [reply ukis_pysat/data.py:268]: We will remove the ProjectionExtension everywhere now to be consistent, can be read from the files.

**Pull request merged**